### PR TITLE
Add maximum memory feature

### PR DESCRIPTION
**Pull request by djsmith42**
This allows callers to configure Workers to respawn if they
consume too much memory.

### CODE REVIEW

**Comment by nebstrebor** [comment]
Thanks Dave.

This feature would be very useful in a memory constrained environment like Heroku, particularly in light of this problem:   https://groups.google.com/d/msg/celery-users/jVc3I3kPtlw/fPfnoP_DhuoJ Heroku seems to be the type of environment Solem describes in his post ("you pay for memory usage")  in that it seems not to be a good fit for that the default sort of memory management where swap is OK and is just part of the game. As soon as things go into swap my dynos slow down and heroku's alarms start going off, and so I have to use `CELERYD_MAX_TASKS_PER_CHILD` as a dumb way to control the memory issue.  The method developed in this PR seems like it would be much smarter and not require unnecessary delay for unnecessary worker restarts.  

Any thoughts from Celery devs on why this PR wouldn't be a useful addition?

**Comment by joshowen** [comment]
This is extremely helpful on providers like Heroku.

**Comment by ask** [comment]
It's a great feature!  Sadly as far as I know getrusage is not actually able to limit RSS memory in any useful way on most platforms, e.g. see http://stackoverflow.com/questions/3043709/resident-set-size-rss-limit-has-no-effect/6365534#6365534

**Comment by djsmith42** [comment]
Ask, after reading that SO answer, I'm not sure how it relates to this PR. It seems to be talking about getrlimit, which my PR does not use. My PR uses getrusage, which does something entirely different.

**Comment by joshowen** [comment]
@ask any more thoughts on this?  It would be VERY useful in memory constrained providers like Heroku

File: billiard/pool.py [hunk 1]
```diff
@@ -241,11 +241,12 @@ class Worker(Process):
 
     def __init__(self, inq, outq, synq=None, initializer=None, initargs=(),
                  maxtasks=None, sentinel=None, on_exit=None,
-                 sigprotection=True):
+                 sigprotection=True, max_mem_kb=None):
```
Comment (ask) [inline billiard/pool.py]: can we use maxmemperchild here? Having kb in the name is useful, but does make the naming kind of awkward :(

**Pull request merged**